File: frontend/src/services/document.service.ts
import api from './api'
import type {
  ApiResponse,
  PaginatedResponse,
} from '../types'
import type {
  DocumentDto,
  UploadResponseDto,
  ReindexResponseDto,
  DocumentContentDto,
  DocumentStatus,
} from '../types/knowledge-base'

/**
 * Progress callback for file upload
 */
export type UploadProgressCallback = (progress: number) => void

/**
 * Document Service
 * Handles all document-related API operations
 */
export const documentService = {
  /**
   * Upload a document
   * @param file - The file to upload
   * @param title - Optional document title
   * @param description - Optional document description
   * @param fileMd5 - Optional file MD5 for instant upload
   * @param onProgress - Optional progress callback
   * @returns Upload response with document info
   */
  async uploadDocument(
    file: File,
    title?: string,
    description?: string,
    fileMd5?: string,
    onProgress?: UploadProgressCallback
  ): Promise<ApiResponse<UploadResponseDto>> {
    try {
      const formData = new FormData()
      formData.append('file', file)
      if (title) formData.append('title', title)
      if (description) formData.append('description', description)
      if (fileMd5) formData.append('fileMd5', fileMd5)

      const response = await api.post<ApiResponse<UploadResponseDto>>(
        '/documents/upload',
        formData,
        {
          headers: {
            'Content-Type': 'multipart/form-data',
          },
          onUploadProgress: (progressEvent) => {
            if (onProgress && progressEvent.total) {
              const progress = Math.round(
                (progressEvent.loaded * 100) / progressEvent.total
              )
              onProgress(progress)
            }
          },
        }
      )

      return response.data
    } catch (error) {
      console.error('Document upload failed:', error)
      throw error
    }
  },

  /**
   * Get documents list with pagination
   * @param page - Page number (0-indexed)
   * @param size - Page size
   * @returns Paginated document list
   */
  async getDocuments(
    page: number = 0,
    size: number = 10
  ): Promise<ApiResponse<PaginatedResponse<DocumentDto>>> {
    try {
      const response = await api.get<ApiResponse<PaginatedResponse<DocumentDto>>>(
        '/documents',
        { params: { page, size } }
      )
      return response.data
    } catch (error) {
      console.error('Failed to get documents:', error)
      throw error
    }
  },

  /**
   * Get current user's documents (alias endpoint)
   * @param page - Page number (0-indexed)
   * @param size - Page size
   * @returns Paginated document list
   */
  async getMyDocuments(
    page: number = 0,
    size: number = 10
  ): Promise<ApiResponse<PaginatedResponse<DocumentDto>>> {
    try {
      const response = await api.get<ApiResponse<PaginatedResponse<DocumentDto>>>(
        '/documents/my',
        { params: { page, size } }
      )
      return response.data
    } catch (error) {
      console.error('Failed to get my documents:', error)
      throw error
    }
  },

  /**
   * Get document by ID
   * @param documentId - Document ID
   * @returns Document details
   */
  async getDocument(documentId: number): Promise<ApiResponse<DocumentDto>> {
    try {
      const response = await api.get<ApiResponse<DocumentDto>>(
        `/documents/${documentId}`
      )
      return response.data
    } catch (error) {
      console.error(`Failed to get document ${documentId}:`, error)
      throw error
    }
  },

  /**
   * Search documents by keyword
   * @param keyword - Search keyword
   * @param page - Page number (0-indexed)
   * @param size - Page size
   * @returns Paginated search results
   */
  async searchDocuments(
    keyword: string,
    page: number = 0,
    size: number = 10
  ): Promise<ApiResponse<PaginatedResponse<DocumentDto>>> {
    try {
      const response = await api.get<ApiResponse<PaginatedResponse<DocumentDto>>>(
        '/documents/search',
        { params: { keyword, page, size } }
      )
      return response.data
    } catch (error) {
      console.error('Failed to search documents:', error)
      throw error
    }
  },

  /**
   * Delete document by ID
   * @param documentId - Document ID
   * @returns Success message
   */
  async deleteDocument(documentId: number): Promise<ApiResponse<string>> {
    try {
      const response = await api.delete<ApiResponse<string>>(
        `/documents/${documentId}`
      )
      return response.data
    } catch (error) {
      console.error(`Failed to delete document ${documentId}:`, error)
      throw error
    }
  },

  /**
   * Update document status
   * @param documentId - Document ID
   * @param status - New status
   * @returns Success message
   */
  async updateDocumentStatus(
    documentId: number,
    status: DocumentStatus
  ): Promise<ApiResponse<string>> {
    try {
      const response = await api.put<ApiResponse<string>>(
        `/documents/${documentId}/status`,
        null,
        { params: { status } }
      )
      return response.data
    } catch (error) {
      console.error(`Failed to update document status ${documentId}:`, error)
      throw error
    }
  },

  /**
   * Download document
   * @param documentId - Document ID
   * @returns Blob with file content
   */
  async downloadDocument(documentId: number): Promise<Blob> {
    try {
      const response = await api.get<Blob>(
        `/documents/${documentId}/download`,
        { responseType: 'blob' }
      )
      return response.data
    } catch (error) {
      console.error(`Failed to download document ${documentId}:`, error)
      throw error
    }
  },

  /**
   * Preview document (for browser-displayable files)
   * @param documentId - Document ID
   * @returns Blob with file content
   */
  async previewDocument(documentId: number): Promise<Blob> {
    try {
      const response = await api.get<Blob>(
        `/documents/${documentId}/preview`,
        { responseType: 'blob' }
      )
      return response.data
    } catch (error) {
      console.error(`Failed to preview document ${documentId}:`, error)
      throw error
    }
  },

  /**
   * Get document text content
   * @param documentId - Document ID
   * @returns Document content metadata
   */
  async getDocumentContent(documentId: number): Promise<ApiResponse<DocumentContentDto>> {
    try {
      const response = await api.get<ApiResponse<DocumentContentDto>>(
        `/documents/${documentId}/content`
      )
      return response.data
    } catch (error) {
      console.error(`Failed to get document content ${documentId}:`, error)
      throw error
    }
  },

  /**
   * Reindex all documents
   * @returns Reindex summary
   */
  async reindexAllDocuments(): Promise<ApiResponse<ReindexResponseDto>> {
    try {
      const response = await api.post<ApiResponse<ReindexResponseDto>>(
        '/documents/reindex'
      )
      return response.data
    } catch (error) {
      console.error('Failed to reindex all documents:', error)
      throw error
    }
  },

  /**
   * Reindex a single document
   * @param documentId - Document ID
   * @returns Success message
   */
  async reindexDocument(documentId: number): Promise<ApiResponse<string>> {
    try {
      const response = await api.post<ApiResponse<string>>(
        `/documents/${documentId}/reindex`
      )
      return response.data
    } catch (error) {
      console.error(`Failed to reindex document ${documentId}:`, error)
      throw error
    }
  },
}

export default documentService

File: backend/services/knowledge-mentor/km-domain/src/main/java/com/company/kb/core/service/DocumentService.java
package com.company.kb.core.service;

import com.company.kb.core.domain.Document;
import com.company.kb.core.domain.DocumentChunk;
import org.springframework.core.io.Resource;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

/**
 * 文档服务接口
 */
public interface DocumentService {

    /**
     * 上传文档
     * @param file 文件
     * @param title 标题
     * @param description 描述
     * @param username 用户名
     * @param fileMd5 文件MD5（用于秒传检查）
     * @return 文档
     */
    Document uploadDocument(MultipartFile file, String title, String description, String username, String fileMd5) throws Exception;

    /**
     * 查找已存在的文档（用于秒传）
     * @param fileMd5 文件MD5
     * @param username 用户名
     * @return 已存在的文档，如果不存在返回null
     */
    Document findExistingDocument(String fileMd5, String username);

    /**
     * 文档上传后授予权限
     * @param documentId 文档ID
     * @param username 用户名
     */
    void grantDocumentPermissionAfterUpload(Long documentId, String username);

    /**
     * 解析文档
     * @param document 文档
     * @return 文档内容
     */
    String parseDocument(Document document) throws Exception;

    /**
     * 分块处理文档
     * @param content 内容
     * @param document 文档
     * @return 文档块列表
     */
    List<DocumentChunk> chunkDocument(String content, Document document);

    /**
     * 索引文档
     * @param document 文档
     * @param chunks 文档块
     */
    void indexDocument(Document document, List<DocumentChunk> chunks) throws Exception;

    /**
     * 删除文档
     * @param documentId 文档ID
     */
    void deleteDocument(Long documentId) throws Exception;

    /**
     * 获取文档
     * @param documentId 文档ID
     * @return 文档
     */
    Document getDocumentById(Long documentId);

    /**
     * 搜索文档
     * @param keyword 关键词
     * @param pageable 分页
     * @return 文档列表
     */
    Page<Document> searchDocuments(String keyword, Pageable pageable);

    /**
     * 获取用户的文档列表
     * @param userId 用户ID
     * @param pageable 分页
     * @return 文档列表
     */
    Page<Document> getUserDocuments(String userId, Pageable pageable);

    /**
     * 更新文档状态
     * @param documentId 文档ID
     * @param status 状态
     */
    void updateDocumentStatus(Long documentId, Document.DocumentStatus status);

    /**
     * 下载文档
     * @param documentId 文档ID
     * @return 下载结果（包含文件名和资源）
     * @throws Exception 下载失败时抛出异常
     */
    FileDownloadResult downloadDocument(Long documentId) throws Exception;

    /**
     * 获取文档的文本内容
     * @param documentId 文档ID
     * @return 文本内容
     * @throws Exception 获取失败时抛出异常
     */
    String getDocumentTextContent(Long documentId) throws Exception;

    /**
     * 获取所有文档（用于批量重新索引）
     * @param pageable 分页
     * @return 文档列表
     */
    Page<Document> getAllDocuments(Pageable pageable);

    /**
     * 获取文档的分块内容
     * @param documentId 文档ID
     * @param pageable 分页
     * @return 文档块分页
     */
    Page<DocumentChunk> getDocumentChunks(Long documentId, Pageable pageable);

    /**
     * 文件下载结果
     */
    class FileDownloadResult {
        private final String fileName;
        private final Resource resource;

        public FileDownloadResult(String fileName, Resource resource) {
            this.fileName = fileName;
            this.resource = resource;
        }

        public String getFileName() {
            return fileName;
        }

        public Resource getResource() {
            return resource;
        }
    }
}
